README.rst
metfilelib
==========================================

This is a library for working with MET files as read by IRIS.

A MET file is represented by the metfilelib.metfile.MetFile
class. Class instances are immutable.

To read in a MET file, use the metfilelib.parser.parse_metfile()
function.  The file_object argument can be created using
metfilelib.util.file_read.FileReader. By default, errors go to stderr,
but you can pass your own error handler to the FileReader.

To write a MET file, use metfilelib.exporters.MetfileExporter.


CHANGES.rst
Changelog of metfilelib
===================================================


0.19 (unreleased)
-----------------

- Nothing changed yet.


0.18 (2019-01-11)
-----------------

- Turns out it had to be drawing type 200.


0.17 (2018-07-24)
-----------------

- Added profile point type 200 (riet).


0.16 (2015-07-08)
-----------------

- Constrain allowed characters in meting, profiel and reeks.


0.15 (2014-03-06)
-----------------

- Retrieving a profile now also ignores empty lines.


0.14 (2014-01-14)
-----------------

- Add 'start_point' property to Profile.

- Add 'as_wkt' property to Point.


0.13 (2013-09-16)
-----------------

- Add missing drawing point code '47'.


0.12 (2013-05-07)
-----------------

- Moved DXF generation from lizard_progress to here.


0.11 (2013-04-23)
-----------------

- Added a helper function to linear_algebra that calculates the
  distance of a point to a line.

- Measurement.point now returns None if something is wrong with one of
  the coordinates.


0.10 (2013-04-22)
-----------------

- Fixes bug with infinite loop in case of wrong date format.


0.9 (2013-04-16)
----------------

- Fixed bug where date format error message was shown on the wrong
  line (lizard-progress #66).

- Added test for #69, but it's working correctly. Perhaps an old
  version?

- Export to MET file used a wrong date format, fixed.


0.8 (2013-04-05)
----------------

- Add lists of all the existing profile point codes and drawing codes,
  check for them.


0.7 (2013-03-29)
----------------

- Make it possible to generate a MET file that sorts its measurements
  based on the projection on the base line.


0.6 (2013-03-27)
----------------

- Add a waterlevel property to Profile, that checks the z1/z2 values at the
  22 profile point codes.

- Improved error handling, we got infinite loops in some cases, avoid them.

- Line now has midpoint and length properties.

- Line has a function distance_to_midpoint, which projects points on the line then
  checks how far from the midpoint it is, with the side of the start point getting
  negative numbers. Useful as X values on plots.

0.5 (2013-03-22)
----------------

- Add version codes for all parser errors.

- Add a method distance() to Point that calculates distance to another
  point.


0.4 (2013-03-21)
----------------

- Check on version number 1.0.

- Try to continue parsing if the second comma in a <REEKS> element is
  missing.

- Fix some bugs in parsing

- Added tools that retrieve some profile from a MET file, and can recreate
  an entire MetFile object from such retrieved profiles

- Added tools for projection X, Y points from measurements, projecting them on
  a base line, and sorting measurements based on their projection on the line.


0.3 (2013-03-19)
----------------

- Translate coordinates to floats.

- Add tests.


0.2 (2013-03-15)
----------------

- Increased the number of error messages. Only error messages to do
  with the correctness of the file are given, checking that the
  content is up to some spec is left to client code.

- See error_codes.rst in the docs/ dir for a list of error codes.


0.1 (2013-03-05)
----------------

- Initial project structure created with nensskel 1.31.dev0.

- Added a util.file_reader.FileReader class that will probably be used
  by the site, or some object with the same API. The metfile parser
  should assume it gets instances of this class.

- Developed metfile.py, parser.parse_metfile and
  exporters.MetfileExporter to the point where a correct MET file can
  be read in and printed out.

- Developed a bit further so that a file can be uploaded now, but still
  definitely work in progress.


metfilelib/metfile.py
# (c) Nelen & Schuurmans.  GPL licensed, see LICENSE.rst.
# -*- coding: utf-8 -*-

"""Python classes that represent the data in a .met file."""

# Python 3 is coming
from __future__ import unicode_literals
from __future__ import print_function
from __future__ import absolute_import
from __future__ import division

from collections import namedtuple

from metfilelib.util import linear_algebra


MetFile = namedtuple('MetFile', 'version, series')

Series = namedtuple('Series', 'line_number, id, name, profiles')


class Profile(namedtuple(
    'Profile',
    '''line_number, id, description, date_measurement, level_value, level_type,
    coordinate_type, number_of_z_values, profile_type_placing,
    start_x, start_y, measurements''')):

    @property
    def line(self):
        """Return the base line of this profile. We use the line
        through the two mandatory '22' profile point types."""
        twentytwos = [m for m in self.measurements
                      if m.profile_point_type == '22']

        if len(twentytwos) != 2:
            return None

        start = twentytwos[0].point
        end = twentytwos[1].point

        if start == end:
            return None

        return linear_algebra.Line(start=start, end=end)

    @property
    def waterlevel(self):
        """Return the water level.

        If there are exactly two '22' profile point types, and their
        z1 and z2 values are all equal, then that's the water
        level. Otherwise return None."""

        twentytwos = [m for m in self.measurements
                      if m.profile_point_type == '22']

        if len(twentytwos) != 2:
            return None

        start = twentytwos[0]
        end = twentytwos[1]

        if start.z1 == start.z2 == end.z1 == end.z2:
            return start.z1

        return None

    @property
    def midpoint(self):
        """Point in the middle of the base line"""
        line = self.line
        if line is None:
            return None

        # Say line is A to B
        # Midpoint is ((B-A)*0.5)+A
        return line.midpoint

    @property
    def sorted_measurements(self):
        """Project points on the base line, return them in order."""
        line = self.line
        if line is None:
            # We can't project without line
            return self.measurements

        return sorted(
            self.measurements,
            key=lambda measurement: (
                line.scaled_scalar_projection(measurement.point)))

    @property
    def water_measurements(self):
        """Return the 22 codes and the measurements in between
        (sorted)"""
        water_indices = []
        measurements = list(self.sorted_measurements)

        for i, measurement in enumerate(measurements):
            if measurement.profile_point_type == '22':
                water_indices.append(i)

        if len(water_indices) != 2:
            # Can't do this
            return None

        return measurements[water_indices[0]:water_indices[1] + 1]

    @property
    def start_point(self):
        return linear_algebra.Point(x=self.start_x, y=self.start_y)


class Measurement(namedtuple(
   'Measurement',
   'line_number, profile_point_type, profile_point_drawing_code, x, y, z1, z2'
        )):
    ALLOWED_PROFILE_POINT_TYPES = set(
        ('1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 '
         '22 98 99').split())

    ALLOWED_DRAWING_CODES = set(
        ('5 10 15 17 19 23 24 25 27 29 30 32 34 35 37 39 41 43 45 47 49 '
         '50 52 54 56 58 60 62 64 66 68 70 72 74 76 78 80 82 84 86 '
         '88 90 92 94 96 98 130 135 140 150 200 999').split())

    @property
    def point(self):
        if not isinstance(self.x, float) or not isinstance(self.y, float):
            return None

        return linear_algebra.Point(x=self.x, y=self.y)
